### ISSUE TITLE
Should parallel version supercede standard version?

### ISSUE THREAD

**Issue by davepl** [PlummersSoftwareLLC/Primes#108]
In the parallel multicore version (in PrimeCPP_PAR) I've added a number of command line switches that allow you to control how many threads, upper limit, how long to run for, and so on.

Since -t is threadcount running "primes.exe -t 1" will run a single-core pass, obviating the need for both a single and multicore versions.

I'm proposing that we orphan the original single core version and that future improvements go into the PAR version as well.

PS: I didn't want to open a separate issue just to gush, but wanted to say I was blown away and thrilled at the response to this project in general and how so many folks have contributed!  And a HUGE thanks to Rolf for stepping up and taking charge!

**Comment by mike-barber** [comment]
I personally think it's a great idea! Nothing is fun until it's multithreaded :) and I like the flexibility with the command line parameters. And we're not really losing anything because you've still got the single-thread too. 

I'm having a play around with a rust version to do the same thing, and I reckon the C# guys will have fun with it too. Really nice to have as an option, for sure! 

For the multithread case, there's a lot of synchronization that I reckon you can eliminate. Essentially we're starting and stopping threads on every iteration, so each calculation has to happen in lockstep. As an alternative, we could just start N threads and let them run independently until the given time expires, then collect the results at the end: i.e. they each watch the time and terminate when appropriate. I played around with the code a little on my machine, and this gives a nice boost. Don't trust my C++ though: I'm a bit out of practice!

```c++
    if (!bOneshot)
    {
        vector<thread> threadPool;
        // pre-allocated place to put results; would probably be more fun with modern promises etc, but
        // this has the benefit of simplicity.
        vector<uint64_t> localCounts(cThreads, 0);
        for (unsigned int i = 0; i < cThreads; i++)
        {
            size_t threadId = i;
            threadPool.push_back(thread([llUpperLimit, tStart, cSeconds, threadId, &localCounts] {
                // run threads until time expires
                auto localPasses = 0;
                while (duration_cast<seconds>(steady_clock::now() - tStart).count() < cSeconds)
                {
                    std::unique_ptr<prime_sieve>(new prime_sieve(llUpperLimit))->runSieve();
                    localPasses++;
                }
                // record final count in localCounts - lock should not be required as we're 
                // updating the elements, not the collection. 
                localCounts[threadId] = localPasses;
            }));
        }

        for (auto &th : threadPool)
            th.join();

        for (auto &c : localCounts)
            cPasses += c;
    }
```

**Comment by Kinematics** [comment]
The main issue I see is confusion about what you mean by "parallel version".

When I first saw a parallel version implementation, I thought you were parallelizing the algorithm itself. I found that interesting because I was having trouble doing that without introducing overhead from locking mechanisms which made it far worse than the linear method.

However you're not making the algorithm parallel, just the running of the algorithm.  Even then, though, there's two different approaches: Are you spinning off an instance of the algorithm that runs in parallel, and only lasts for one instance of the algorithm run, or are you creating threads that keep running the algorithm until the time is up?

So basically, there are three different things that you can mean by calling it "parallel", and the term itself doesn't explain what you're doing.  That then introduces potential confusion in how the other language implementations might create comparable versions.

I would suggest instead that you have the algorithm class in its own class file (as proper coding standards would normally suggest), and have the runner (eg: main()) in an entirely separate file, self-contained, and have it call a descriptive function that runs things as desired, rather than bundling everything inside main().

EG (pseudocode):

```
main(string[] args) {
    myArgs = ParseCommandline(args);

    if (myArgs.RunMultipleThreads)
        RunMultipleThreads();
    else if (myArgs.RunThreadPerInstance)
        RunThreadPerInstance();
    else if (myArgs.RunParallelized)
        RunParallelizedAlgorithm();
    else
        Run();
}

RunMultipleThreads() {
    // Create X threads and let each one repeatedly run the algorithm until time is up
}

RunThreadPerInstance() {
    // Run each algorithm instance on its own thread
}

RunParallelizedAlgorithm() {
    // Run a version of the algorithm that is internally parallelized
}

Run() {
    // Run the standard algorithm until time is up
}
```

(Speaking of args[], don't roll your own. Get a library that handles commandline argument parsing.)

**Comment by mike-barber** [comment]
> Are you spinning off an instance of the algorithm that runs in parallel, and only lasts for one instance of the algorithm run, or are you creating threads that keep running the algorithm until the time is up?

I think I'm responsible for this part of the confusion :) Dave was doing the former, and I was suggesting the latter in my post above. 

I'm not sure what the best way to benchmark this stuff is, as there's a fair bit of variation between languages. I suspect the more "embarrassingly parallel" the approach is, the more widely it can be used.

**Comment by rhbvkleef** [comment]
It looks like we'll keep the parallel implementation separate. The discussion about this is taking place in #110 .

**Comment by rbergen** [comment]
I think we can close this. @rhbvkleef , @marghidanu ?

**Comment by marghidanu** [comment]
Yep!
